File: src/main/java/com/cdot/lists/ChecklistActivity.java
/*
 * Copyright © 2020 C-Dot Consultants
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.cdot.lists;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.InputType;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.FileProvider;

import com.cdot.lists.databinding.ChecklistActivityBinding;
import com.cdot.lists.model.Checklist;
import com.cdot.lists.model.ChecklistItem;
import com.cdot.lists.model.EntryList;
import com.cdot.lists.model.EntryListItem;
import com.cdot.lists.preferences.PreferencesActivity;
import com.cdot.lists.view.ChecklistItemView;
import com.cdot.lists.view.EntryListItemView;
import com.opencsv.CSVWriter;

import java.io.File;
import java.io.FileWriter;
import java.io.Writer;
import java.util.List;

/**
 * Activity for managing interactions with a checklist of items
 */
public class ChecklistActivity extends EntryListActivity {
    private static final String TAG = ChecklistActivity.class.getSimpleName();
    // When in edit mode, sorting and moving checked items is disabled
    public boolean mInEditMode = false;
    // helper to avoid frequent casts
    private Checklist mChecklist;
    private ChecklistActivityBinding mBinding;

    public ChecklistActivity() {
    }

    @Override
        // EntryListActivity
    EntryList getList() {
        return mChecklist;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle state) {
        super.onSaveInstanceState(state);
        state.putInt(UID_EXTRA, getList().getSessionUID());
    }

    @Override
    public void onRestoreInstanceState(@NonNull Bundle state) {
        super.onRestoreInstanceState(state);
        if (mChecklist == null) {
            int uid = state.getInt(UID_EXTRA);
            if (uid > 0)
                mChecklist = (Checklist) getLister().getLists().findBySessionUID(uid);
        }
    }

    @Override // AppCompatActivity
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = getIntent();
        int uid = intent.getIntExtra(UID_EXTRA, -1);
        EntryList lists = getLister().getLists();
        mChecklist = (Checklist) lists.findBySessionUID(uid);

        mBinding = ChecklistActivityBinding.inflate(getLayoutInflater());

        makeAdapter(mBinding.itemListView);

        mBinding.addItemET.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_FLAG_AUTO_COMPLETE | InputType.TYPE_TEXT_FLAG_CAP_SENTENCES);
        mBinding.addItemET.setOnEditorActionListener((textView, i, keyEvent) -> {
            if (i == EditorInfo.IME_ACTION_DONE) {
                addNewItem();
                return true;
            }
            return false;
        });

        mBinding.addItemET.setImeOptions(EditorInfo.IME_ACTION_DONE);

        enableEditMode(getList().size() == 0);

        setContentView(mBinding.getRoot());
    }

    @Override // implements EntryListActivity
    protected String getHelpAsset() {
        return "Checklist";
    }

    @Override // implements EntryListActivity
    protected EntryListItemView makeMovingView(EntryListItem item) {
        return new ChecklistItemView(item, true, this);
    }

    @Override // Activity
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        Log.d(TAG, "onCreateOptionsMenu");
        getMenuInflater().inflate(R.menu.checklist, menu);
        return true;
    }

    @Override // EntryListActivity
    public boolean onPrepareOptionsMenu(@NonNull Menu menu) {
        Log.d(TAG, "onPrepareOptionsMenu");
        super.onPrepareOptionsMenu(menu);

        MenuItem it = menu.findItem(R.id.action_edit);

        if (mInEditMode) {
            it.setIcon(R.drawable.ic_action_item_add_off);
            it.setTitle(R.string.action_item_add_off);
        } else {
            it.setIcon(R.drawable.ic_action_item_add_on);
            it.setTitle(R.string.action_item_add_on);
        }

        menu.findItem(R.id.action_settings).setEnabled(!mInEditMode);
        menu.findItem(R.id.action_check_all).setEnabled(((Checklist) getList()).getCheckedCount() < getList().size());
        menu.findItem(R.id.action_uncheck_all).setEnabled(((Checklist) getList()).getCheckedCount() > 0);
        menu.findItem(R.id.action_undo_delete).setEnabled(getList().getRemoveCount() > 0);
        menu.findItem(R.id.action_delete_checked).setEnabled(((Checklist) getList()).getCheckedCount() > 0);

        return true;
    }

    @Override // EntryListActivity
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        // Beware dispatchTouchEvent stealing events
        Log.d(TAG, "onOptionsItemSelected");
        if (super.onOptionsItemSelected(menuItem))
            return true;

        int it = menuItem.getItemId();

        if (it == R.id.action_check_all) {
            if (mChecklist.checkAll(true)) {
                getList().notifyChangeListeners();
                Log.d(TAG, "check all");
                checkpoint();
            }

        } else if (it == R.id.action_delete_checked) {
            int deleted = mChecklist.deleteAllChecked();
            if (deleted > 0) {
                mChecklist.notifyChangeListeners();
                Log.d(TAG, "checked deleted");
                checkpoint();
                Toast.makeText(this, getString(R.string.items_deleted, deleted), Toast.LENGTH_SHORT).show();
                if (getList().size() == 0) {
                    enableEditMode(true);
                    mChecklist.notifyChangeListeners();
                    checkpoint();
                }
            }

        } else if (it == R.id.action_edit)
            enableEditMode(!mInEditMode);

        else if (it == R.id.action_rename_list) {
            AlertDialog.Builder builder = new androidx.appcompat.app.AlertDialog.Builder(this);
            builder.setTitle(R.string.rename_list);
            builder.setMessage(R.string.enter_new_name_of_list);
            final EditText editText = new EditText(this);
            editText.setSingleLine(true);
            editText.setText(getList().getText());
            builder.setView(editText);
            builder.setPositiveButton(R.string.ok, (dialogInterface, i) -> {
                Log.d(TAG, "list renamed");
                getList().setText(editText.getText().toString());
                getLister().notifyListsListeners();
                checkpoint();
                //Objects.requireNonNull(getSupportActionBar()).setTitle(mChecklist.getText());
            });
            builder.setNegativeButton(R.string.cancel, null);
            builder.show();

        } else if (it == R.id.action_uncheck_all) {
            if (mChecklist.checkAll(false)) {
                Log.d(TAG, "uncheck all");
                getList().notifyChangeListeners();
                checkpoint();
            }

        } else if (it == R.id.action_undo_delete) {
            int undone = mChecklist.undoRemove();
            if (undone == 0)
                Toast.makeText(this, R.string.no_deleted_items, Toast.LENGTH_SHORT).show();
            else {
                mChecklist.notifyChangeListeners();
                Log.d(TAG, "delete undone");
                checkpoint();
                Toast.makeText(this, getString(R.string.items_restored, undone), Toast.LENGTH_SHORT).show();
            }

        } else if (it == R.id.action_save_list_as)
            exportChecklist();

        else if (it == R.id.action_settings) {
            Intent intent = new Intent(this, PreferencesActivity.class);
            intent.putExtra(UID_EXTRA, mChecklist.getSessionUID());
            startActivityForResult(intent, REQUEST_PREFERENCES);

        } else
            return super.onOptionsItemSelected(menuItem);

        return true;
    }

    @Override // EntryListActivity
    public boolean canManualSort() {
        return !mChecklist.getFlag(Checklist.moveCheckedItemsToEnd) && super.canManualSort();
    }

    @Override // EntryListActivity
    protected List<EntryListItem> getDisplayOrder() {
        if (mInEditMode)
            return getList().getData();

        List<EntryListItem> list = super.getDisplayOrder(); // get sorted list
        if (mChecklist.getFlag(Checklist.moveCheckedItemsToEnd)) {
            int top = list.size();
            int i = 0;
            while (i < top) {
                ChecklistItem item = (ChecklistItem) list.get(i);
                if (item.getFlag(ChecklistItem.isDone)) {
                    list.add(list.remove(i));
                    top--;
                } else
                    i++;
            }
        }
        return list;
    }

    /**
     * Turn editing "mode" on/off
     *
     * @param isOn new state
     */
    private void enableEditMode(boolean isOn) {
        mInEditMode = isOn;
        notifyAdapter();
        invalidateOptionsMenu();

        // Show/hide soft keyboard
        InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        if (mInEditMode) {
            mBinding.addItemET.setVisibility(View.VISIBLE);
            mBinding.addItemET.setFocusable(true);
            mBinding.addItemET.setFocusableInTouchMode(true);
            mBinding.addItemET.requestFocus();
            inputMethodManager.showSoftInput(mBinding.addItemET, InputMethodManager.SHOW_IMPLICIT);
        } else {
            mBinding.addItemET.setVisibility(View.INVISIBLE);
            View currentFocus = getCurrentFocus();
            if (currentFocus == null)
                currentFocus = mBinding.addItemET;
            inputMethodManager.hideSoftInputFromWindow(currentFocus.getWindowToken(), 0);
        }
    }

    /**
     * Handle adding an item from the typing area
     */
    private void addNewItem() {
        String text = mBinding.addItemET.getText().toString();
        if (text.trim().length() == 0)
            return;
        EntryListItem find = getList().findByText(text, false);
        if (find == null || !mChecklist.getFlag(EntryList.warnAboutDuplicates))
            addItem(text);
        else
            promptSimilarItem(text, find.getText());
    }

    /**
     * Handle adding an item after it's confirmed
     *
     * @param str the text of the item
     */
    private void addItem(String str) {
        ChecklistItem item = new ChecklistItem(mChecklist, str, false);
        getList().add(item);
        Log.d(TAG, "item added");
        getList().notifyChangeListeners();
        mBinding.addItemET.setText("");
        mBinding.itemListView.smoothScrollToPosition(getDisplayOrder().indexOf(item));
        checkpoint();
    }

    /**
     * Prompt for similar item already in the list
     *
     * @param proposed the text of the proposed item
     * @param similar  the text of a similar item already in the list
     */
    private void promptSimilarItem(final String proposed, String similar) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.similar_item_already_in_list);
        builder.setMessage(getString(R.string.similar_item_x_already_in_list, similar, proposed));
        builder.setPositiveButton(R.string.ok, (dialogInterface, i) -> addItem(proposed));
        builder.setNegativeButton(R.string.cancel, null);
        builder.show();
    }

    /**
     * Export the checklist in a user-selected format
     */
    private void exportChecklist() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.export_format);
        final Spinner picker = new Spinner(this);
        // Helper for indexing the array of share formats, must be final for inner class access
        final int[] mPlace = new int[1];
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_spinner_item, getResources().getStringArray(R.array.share_format_description));
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        picker.setAdapter(adapter);
        picker.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent, View v, int position, long id) {
                mPlace[0] = position;
            }

            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
        builder.setView(picker);

        final String listName = getList().getText();

        builder.setPositiveButton(R.string.ok, (dialog, id) -> {
            final Intent intent = new Intent(Intent.ACTION_SEND);

            intent.putExtra(Intent.EXTRA_TITLE, listName); // Dialog title

            String mimeType = getResources().getStringArray(R.array.share_format_mimetype)[mPlace[0]];
            intent.setType(mimeType);

            String ext = getResources().getStringArray(R.array.share_format_mimeext)[mPlace[0]];
            String fileName = listName.replaceAll("[/\u0000]", "_") + ext;
            // WTF! The EXTRA_SUBJECT is used as the document title for Drive saves!
            intent.putExtra(Intent.EXTRA_SUBJECT, fileName);

            // text body e.g. for email
            String text = mChecklist.toPlainString("");
            intent.putExtra(Intent.EXTRA_TEXT, text);

            try {
                // Write a local file for the attachment
                File sendFile = new File(getExternalFilesDir("send"), fileName);
                Writer w = new FileWriter(sendFile);
                switch (mimeType) {
                    case "text/plain":
                        // Re-write the text body in the attachment
                        w.write(text);
                        break;

                    case "application/json":
                        w.write(mChecklist.toJSON().toString());
                        break;

                    case "text/csv":
                        CSVWriter csvw = new CSVWriter(w);
                        mChecklist.toCSV(csvw);
                        break;

                    default:
                        throw new Exception("Unrecognised share format");
                }
                w.close();

                // Expose the local file using a URI from the FileProvider, and add the URI to the intent
                // See https://medium.com/@ali.muzaffar/what-is-android-os-fileuriexposedexception-and-what-you-can-do-about-it-70b9eb17c6d0
                String authRoot = getPackageName().replace(".debug", "");
                Uri uri = FileProvider.getUriForFile(this, authRoot + ".provider", sendFile);
                intent.putExtra(Intent.EXTRA_STREAM, uri);

                // Fire off the intent
                startActivity(Intent.createChooser(intent, fileName));

            } catch (Exception e) {
                String mess = getString(R.string.failed_export, e.getMessage());
                Log.d(TAG, mess);
                Toast.makeText(this, mess, Toast.LENGTH_SHORT).show();
            }
        });
        builder.setNegativeButton(R.string.cancel, null);
        builder.show();
    }
}

File: src/main/java/com/cdot/lists/ChecklistActivity.kt
/*
 * Copyright © 2020 C-Dot Consultants
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.cdot.lists

import android.content.DialogInterface
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.InputType
import android.util.Log
import android.view.*
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.appcompat.app.AlertDialog
import com.cdot.lists.databinding.ChecklistActivityBinding
import com.cdot.lists.model.*
import com.cdot.lists.preferences.PreferencesActivity
import com.cdot.lists.view.ChecklistItemView
import com.cdot.lists.view.EntryListItemView
import org.json.JSONException

/**
 * Activity for managing interactions with a checklist of items
 */
class ChecklistActivity : EntryListActivity() {
    private lateinit var binding: ChecklistActivityBinding // init in onCreate
    private var isPinned: Boolean = false

    override lateinit var list: EntryList  // init in onCreate

    override fun onSaveInstanceState(state: Bundle) {
        super.onSaveInstanceState(state)
        state.putInt(UID_EXTRA, list.sessionUID)
        state.putInt(PINNED, if (isPinned) 1 else 0)
    }

    override fun onRestoreInstanceState(state: Bundle) {
        super.onRestoreInstanceState(state)
        val uid = state.getInt(UID_EXTRA)
        if (uid > 0) list = lister.lists.findBySessionUID(uid) as EntryList
        isPinned = (state.getInt(PINNED) == 1)
    }

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)
        binding = ChecklistActivityBinding.inflate(layoutInflater)
        val intent = intent
        val uid = intent.getIntExtra(UID_EXTRA, -1)
        val lists: EntryList = lister.lists
        list = lists.findBySessionUID(uid) as EntryList
        Log.d(TAG, "onCreate list $list")
        makeAdapter()
        binding.addItemET.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_AUTO_COMPLETE or InputType.TYPE_TEXT_FLAG_CAP_SENTENCES
        binding.addItemET.setOnEditorActionListener { _: TextView?, i: Int, _: KeyEvent? ->
            if (i == EditorInfo.IME_ACTION_DONE) {
                val text = binding.addItemET.text.toString()
                if (text.trim { it <= ' ' }.isNotEmpty()) {
                    if (list.getFlag(EntryList.WARN_ABOUT_DUPLICATES)) {
                        val found = list.findByText(text, false)
                        if (found == null)
                            addItem(text)
                        else
                            promptSimilarItem(text, found.text)
                    } else
                        addItem(text)
                }
                return@setOnEditorActionListener true
            }
            false
        }
        binding.addItemET.imeOptions = EditorInfo.IME_ACTION_DONE
        if (list.size() == 0) enableAddingMode()
        setContentView(binding.root)
        supportActionBar!!.title = list.text
        pin()
    }

    override val helpAsset: Int = R.raw.checklist_help

    override val listView: ListView
        get() = binding.itemListView

    override val rootView: View
        get() = binding.root

    override val addItemTextView : TextView
        get() = binding.addItemET

    override fun makeItemView(movingItem: EntryListItem, drag: Boolean): EntryListItemView {
        return ChecklistItemView(movingItem, drag, this)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        //Log.d(TAG, "onCreateOptionsMenu");
        menuInflater.inflate(R.menu.checklist, menu)
        return true
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        super.onPrepareOptionsMenu(menu)
        val pinned = menu.findItem(R.id.action_pin)
        if (isPinned) {
            pinned.setIcon(R.drawable.ic_action_unpin)
            pinned.setTitle(R.string.action_unpin)
        } else {
            pinned.setIcon(R.drawable.ic_action_pin)
            pinned.setTitle(R.string.action_pin)
        }

        menu.findItem(R.id.action_preferences).isEnabled = !addItemTextView.isShown
        menu.findItem(R.id.action_check_all).isEnabled = (list as Checklist).countFlaggedEntries(ChecklistItem.IS_DONE) < list.size()
        menu.findItem(R.id.action_uncheck_all).isEnabled = (list as Checklist).countFlaggedEntries(ChecklistItem.IS_DONE) > 0
        menu.findItem(R.id.action_undo_delete).isEnabled = list.removeCount > 0
        menu.findItem(R.id.action_delete_checked).isEnabled = (list as Checklist).countFlaggedEntries(ChecklistItem.IS_DONE) > 0
        return true
    }

    override fun onOptionsItemSelected(menuItem: MenuItem): Boolean {
        // Beware dispatchTouchEvent stealing events
        //Log.d(TAG, "onOptionsItemSelected");
        if (super.onOptionsItemSelected(menuItem)) return true
        val checklist = list as Checklist
        when (menuItem.itemId) {

            R.id.action_pin -> {
                isPinned = !isPinned
                Log.d(TAG, if (isPinned) "Pinning" else "Unpinning")
                pin()
                messageHandler.sendMessage(messageHandler.obtainMessage(MESSAGE_UPDATE_DISPLAY))
            }

            R.id.action_check_all -> if (checklist.setFlagOnAll(ChecklistItem.IS_DONE, true)) {
                list.notifyChangeListeners()
                Log.d(TAG, "check all")
                checkpoint()
            }

            R.id.action_delete_checked -> {
                val deleted = checklist.deleteAllFlagged(ChecklistItem.IS_DONE)
                if (deleted > 0) {
                    list.notifyChangeListeners()
                    Log.d(TAG, "checked deleted")
                    checkpoint()
                    reportShort(R.string.snack_deleted, deleted)
                    if (list.size() == 0) {
                        enableAddingMode()
                        list.notifyChangeListeners()
                        checkpoint()
                    }
                }
            }

            R.id.action_add_items -> toggleAddingMode()

            R.id.action_rename_list -> {
                val builder = AlertDialog.Builder(this)
                builder.setTitle(R.string.rename_list)
                builder.setMessage(R.string.enter_new_name_of_list)
                val editText = EditText(this)
                editText.isSingleLine = true
                editText.setText(list.text)
                builder.setView(editText)
                builder.setPositiveButton(R.string.ok) { _: DialogInterface?, _: Int ->
                    Log.d(TAG, "list renamed")
                    list.text = editText.text.toString()
                    lister.lists.notifyChangeListeners()
                    checkpoint()
                }
                builder.setNegativeButton(R.string.cancel, null)
                builder.show()
            }

            R.id.action_uncheck_all ->
                if (checklist.setFlagOnAll(ChecklistItem.IS_DONE, false)) {
                    Log.d(TAG, "uncheck all")
                    list.notifyChangeListeners()
                    checkpoint()
                }

            R.id.action_undo_delete -> {
                val undone = checklist.undoRemove()
                if (undone == 0) reportShort(R.string.snack_no_deleted_items) else {
                    Log.d(TAG, "delete undone")
                    list.notifyChangeListeners()
                    checkpoint()
                    reportShort(R.string.snack_restored, undone)
                }
            }

            R.id.action_share_list ->
                try {
                    // This doesn't have to be fast, so we use the clunky but safe mechanism of serialising to
                    // JSON to simplify parenthood
                    val container = Checklists()
                    container.addChild(Checklist().fromJSON(list.toJSON()))
                    share(container, list.text)
                } catch (je: JSONException) {
                    Log.e(TAG, Lister.stringifyException(je))
                }

            R.id.action_preferences -> {
                val intent = Intent(this, PreferencesActivity::class.java)
                intent.putExtra(UID_EXTRA, list.sessionUID)
                startActivityForResult(intent, REQUEST_PREFERENCES)
            }

            else -> return super.onOptionsItemSelected(menuItem)
        }
        return true
    }

    override fun canManualSort(): Boolean {
        return !list.getFlag(Checklist.CHECKED_AT_END) && super.canManualSort()
    }

    override fun updateDisplay() {
        // Could simplify to runOnUiThread { notifyAdapter() }
        super.updateDisplay()
        supportActionBar?.title = list.text
    }

    override val displayOrder: MutableList<EntryListItem>
        get() {
            val dl = super.displayOrder // get sorted list
            if (list.getFlag(Checklist.CHECKED_AT_END)) {
                var top = dl.size
                var i = 0
                while (i < top) {
                    val item = dl[i] as ChecklistItem
                    if (item.getFlag(ChecklistItem.IS_DONE)) {
                        dl.add(dl.removeAt(i))
                        top--
                    } else i++
                }
            }
            return dl
        }

    override fun addItem(str: String) {
        val item = ChecklistItem(str)
        list.addChild(item)
        Log.d(TAG, "item $item added to $list")
        binding.addItemET.setText("")
        binding.itemListView.smoothScrollToPosition(displayOrder.indexOf(item))
        checkpoint()
    }

    override fun onAttachedToWindow() {
        // onAttachedToWindow is called after onResume (and it happens only once per lifecycle).
        // ActivityThread.handleResumeActivity call will add DecorView to the current WindowManger
        // which will in turn call WindowManagerGlobal.addView() which than traverse all the views
        // and call onAttachedToWindow on each view.
        pin()
    }

    // Do whatever is needed to keep the app in front of the lock screen
    private fun pin() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1)
            setShowWhenLocked(isPinned)
        else {
            if (isPinned)
                window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
            else
                window.clearFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
        }

        // Do what is needed to keep the device awake while the app is active
        if (isPinned)
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        else
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    companion object {
        private val TAG = ChecklistActivity::class.simpleName
    }
}